apps/src/ni_agast.cpp

```diff
@@ -54,6 +52,10 @@
 #include <pcl/console/print.h>
 #include <pcl/console/parse.h>
 
+#include <boost/thread/mutex.hpp>
+
+#include <thread>
```

Comment:
Is there a particular reason for this order of includes?

Typically I try to arrange my includes in the order of "decreasing library size". That is, standard library headers first, then comprehensive libraries/frameworks like Boost, Qt, VTK, then smaller third-party dependencies (e.g. QHull). Finally, my own headers (PCL in this case).

Reply:
I use the reverse order (compared to you) to pick up potential missing includes. Imagine you have the following headers `<a.h>`,`<b.h>`.
```cpp
// a.h
class A {};
```
```cpp
// b.h

// deliberately forgot to include a.h
class B : public A {};
```

Now imagine that in all files that include `<b.h>`, for some reason, you also always include `<a.h>` first. A strategy not necessarily equivalent but with some similarities to your own. The more elementary classes and definitions are included first.

In this situation your header `<b.h>` is missing an include, but since every time the compiler looks at it, it already parsed `<a.h>` it won't complain that `A` does not exists.

Using my order, this simply can't happen.

Reply:
Hmm, your approach has a point. In fact, now that I'm googling this topic, it seems to be a widely used convention. Perhaps we should adopt it here as well.